### DISCUSSION TITLE
use tokio and actix-web together

### DISCUSSION THREAD

**Discussion by luxl** [tokio-rs/tokio#3025] [Q&A] [answered]
I have a project which needs to have both a http server and a tcp server. So I choose actix-web and tokio(version 0.3.1).
But when I do the following in main.rs (start both tcp server and http server): 
```
    tokio::spawn(async move {
        PetServer::new()
            .bind("0.0.0.0:8898").unwrap()
            .run().await.unwrap();
    });

    HttpServer::new(move || {
        App::new()
            .configure(pet_routes)
    }).bind("0.0.0.0:41310")?.run().await
````

It compiles, but it panic at runtime:
```
thread 'main' panicked at 'must be called from the context of Tokio runtime configured with either `basic_scheduler` or `threaded_scheduler`',
```

If I use the same tokio version as what actix-web depends(which is 0.2.x),  it's ok.
So is this beacuse imcompatible between 0.2 and 0.3? Is there any other solutions if I want to use version 0.3 tokio while using actix-web(which is based on tokio 0.2.x)?

**Comment by Darksonn** [comment, marked as answer]
Tokio 0.2 and 0.3 are not directly compatible, but there is a compatibility layer available, which is described in the [Compatibility between 0.2 and 0.3][1] section of the 0.3 announcement.

Assuming `HttpServer` requires Tokio 0.2, you can use it like this:
```rust
use tokio_compat_02::FutureExt;

start_server().compat().await;
```
```rust
async fn start_server() {
    HttpServer::new(move || {
        App::new()
            .configure(pet_routes)
    }).bind("0.0.0.0:41310")?.run().await
}
```

[1]: https://tokio.rs/blog/2020-10-tokio-0-3#compatibility-between-02-and-03

**Comment by ghost** [comment]
Hi @Darksonn, Do you know if using the compatibility layer induces a performance cost, or if I should rather stay on `tokio = "0.2"`?

**Reply by davidbarsky** [reply]
If there is a cost, it's minimal at best. It's not something I'd worry about.

**Reply by Darksonn** [reply]
The compatibility has the cost of starting an 0.2 runtime alongside your 0.3 runtime. It's probably simpler to just use Tokio 0.2.
